runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/ProcessBundleDescriptors.java

```diff
@@ -146,13 +189,89 @@ private static TargetEncoding addStageOutput(
         wireCoder);
   }
 
+  private static Map<String, Map<String, MultimapSideInputSpec>> forMultimapSideInputs(
+      ExecutableStage stage,
+      Components components,
+      ProcessBundleDescriptor.Builder bundleDescriptorBuilder) throws IOException {
+    ImmutableTable.Builder<String, String, MultimapSideInputSpec> idsToSpec =
+        ImmutableTable.builder();
+    for (SideInputReference sideInputReference : stage.getSideInputs()) {
+      // Update the coder specification for side inputs to be length prefixed so that the
+      // SDK and Runner agree on how to encode/decode the key, window, and values for multimap
+      // side inputs.
+      String pCollectionId = sideInputReference.collection().getId();
+      RunnerApi.MessageWithComponents lengthPrefixedSideInputCoder =
+          LengthPrefixUnknownCoders.forCoder(
+              components.getPcollectionsOrThrow(pCollectionId).getCoderId(),
+              components,
+              false);
+      String wireCoderId =
+          addWireCoder(sideInputReference.collection(), components, bundleDescriptorBuilder);
+      String lengthPrefixedSideInputCoderId = SyntheticComponents.uniqueId(
+          String.format(
+              "fn/side_input/%s",
+              components.getPcollectionsOrThrow(pCollectionId).getCoderId()),
+          bundleDescriptorBuilder.getCodersMap().keySet()::contains);
+
+      bundleDescriptorBuilder.putCoders(
+          lengthPrefixedSideInputCoderId, lengthPrefixedSideInputCoder.getCoder());
+      bundleDescriptorBuilder.putAllCoders(
+          lengthPrefixedSideInputCoder.getComponents().getCodersMap());
+      bundleDescriptorBuilder.putPcollections(
+          pCollectionId,
+          bundleDescriptorBuilder
+              .getPcollectionsMap()
+              .get(pCollectionId)
+              .toBuilder()
+              .setCoderId(lengthPrefixedSideInputCoderId)
+              .build());
+
+      FullWindowedValueCoder<KV<?, ?>> coder =
+          (FullWindowedValueCoder) WireCoders.instantiateRunnerWireCoder(
```

Comment:
The correctness of this change now depends on `instantiateRunnerWireCoder` having the same behavior as the hand coder instantiation above. Can we tie the implementations together somehow? For example, by moving side input SDK coder construction into `WireCoders`? It would also be good to add a comment mentioning that those constructions must be kept in sync.

Reply:
I would rather have WireCoders not be exposed and actually hidden away as an implementation detail of how this specific executable process bundle descriptor handles materialization of side inputs and also how it handles side input access. Other usages should be migrated to either know what they are doing via LengthPrefixUnknownCoders or be modeled as an ExecutableStage and hence can then use the ExecutableStage library code.

Reply:
I would like that in practice, although currently we require WireCoders on the runner side (at least in Flink) in order to give collections an associated serialization mechanism. This may change down the line when everything is "just bytes", but that will require some additional machinery. For example, to extract keys and values for grouping operations.

The good news is that runner-side coders need not be kept in sync with the wire coders here. Anyway, this seems fine for now and we can address that when we get around to it.

Reply:
Serialization will also be needed for push back with side inputs (at least for streaming).

In the old, non-portable Flink runner's ParDo execution, the operator also requires access to the key to manage state and timers. Will the key extraction be in the SDK (and a parameter to the runner)?

Reply:
As the current interfaces are structured, we only support side inputs that don't need pushback (batch pipelines). Another interface will be added once main window -> side input window mapping is added to the `BundleProcessor` for side inputs for unbounded PCollections

Yes, the SDK will be responsible for key extraction and will be provided to the Runner. It will be analogous to how the current MultimapSideInputHandler is provided key, window, and relevant coders. See the BagUserStateHandler part of this PR.

Reply:
Thanks for the clarification. Is there a JIRA that tracks side input support for unbounded collections? I would like to add it as pointer to the corresponding translation code in https://github.com/apache/beam/pull/5407.